File: nutshell/disjointset.cpp
#ifndef DISJOINT_SET
#define DISJOINT_SET 1

#include <iostream>
#include <vector>

using std::cout;
using std::endl;
using std::vector;

namespace nutshell {

class DisjointSet {
public:

    template<class T> void swap(T &A, T &B) {
        T tmp = A;
        A = B;
        B = tmp;
    }

    vector<int> parent;
    vector<int> rank;

    DisjointSet(int n) {
        parent.resize(n);
        for (int i = 0; i < n; ++i) {
            parent[i] = i;
        }
        rank.resize(n, 0);
    }

    int find(int x) {
        if (parent[x] != x) {
            parent[x] = find(parent[x]); // path compression
            return parent[x];
        } else {
            return x;
        }
    }

    bool same(int x, int y) {
        return find(x) == find(y);
    }

    // union by rank
    void link(int x, int y) {
        if (rank[x] > rank[y]) {
            parent[y] = x;
        } else {
            parent[x] = y;
            if (rank[x] == rank[y]) {
                rank[y]++;
            }
        }
    }

    void merge(int x, int y) {
        int rx = find(x);
        int ry = find(y);

        if (rx == ry) {
            return ;
        }

        link(rx, ry);
    }
};

}
#endif // DISJOINT_SET

File: ruby/disjoint_set.rb
class DisjointSet
  attr_reader :leader
  def initialize(n)
    @leader = Array.new n, -1
  end

  def find(x)
    return x if leader[x] < 0
    leader[x] = find(leader[x]) # pass compression
  end

  def size(x)
    -leader[find(x)]
  end

  def same?(x, y)
    find(x) == find(y)
  end

  def link!(x, y)
    x,y=y,x if leader[x] > leader[y]
    leader[x] += leader[y]
    leader[y] = x
  end

  def merge!(x, y)
    rx = find(x)
    ry = find(y)
    return false if rx == ry
    link!(rx, ry)
    true
  end
end
